### ISSUE TITLE
Extreme size and compile time bloat

### ISSUE THREAD

**Issue by arnetheduck** [status-im/nim-chronicles#29]
Currently, every single log line instantiates and inlines a full set of data-to-output formatting code, with essentially no code reuse (due to the use of templates vs proc) - this leads to extreme bloat, both for compile time and code size.

Here are some stats, when logging `info "hello", a=1, b=2, c=3, d=4, e=5` (on a pretty decent laptop):

times|compile-time-debug|compile-time-release|c-file-size-release|release-binary-size
---|----|----|----|----
10|1.7s|3.1s|372k|208k
100|6.4s|1m9s|3572k|1072k

**Comment by kdeme** [comment]
I would suggest to first work on #20. Add a few test cases which measure time and file size just to have a first reference.
Next, once all test cases are implemented, improvements can be done and more easily run against all the tests. And time/size can be referenced against previous tests.

**Comment by arnetheduck** [comment]
Here's the sample I used..

[size.nim.txt](https://github.com/status-im/nim-chronicles/files/2411739/size.nim.txt)

**Comment by zah** [comment]
Fixed in 44b87f8927ec2349163307e4720b27c7c14f297f. Here is the new table (using `--forcebuild`):

times|compile-time-debug|compile-time-release|c-file-size-release|release-binary-size
---|----|----|----|----
10|0.9s|1.7s|19k|131k
100|1.0s|1.8s|143k|157k
1000|1.9s|2.7s|1346k|380k

A single log statement is now compiled to code looking like this:

```
{
  initLogRecord_pKwVKSywZSyAzmyD5sUfmg((&record_zOcHG9ceJivPesM2709bSU9aw), ((tyEnum_LogLevel_9bciZEQ8iC1G2PermjsuXmA) 3), ((NimStringDesc*) NIM_NIL), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_2));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_3), ((NI) 0));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_4), ((NI) 1));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_5), ((NI) 2));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_6), ((NI) 3));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_7), ((NI) 4));
  setProperty_NBa85rQfGg124EfeWfVDzw((&record_zOcHG9ceJivPesM2709bSU9aw), ((NimStringDesc*) &TM_9bJZ0iwMyzUSnudLLDV1lkA_8), ((NI) 5));
  logAllDynamicProperties_zX0QFfzubG1odN00s7I9bnQ((&record_zOcHG9ceJivPesM2709bSU9aw));
  flushRecord_YKZiGmJyWp8Nj1IQPpnm9bw((&record_zOcHG9ceJivPesM2709bSU9aw));
}
```
